compiler/rustc_interface/src/interface.rs

```diff
@@ -124,7 +124,10 @@ pub fn parse_cfgspecs(cfgspecs: Vec<String>) -> FxHashSet<(String, Option<String
                     Err(errs) => errs.into_iter().for_each(|mut err| err.cancel()),
                 }
 
-                error!(r#"expected `key` or `key="value"`"#);
+                error!(concat!(
+                    r#"expected `key` or `key="value"`, ensure escaping is appropriate"#,
+                    r#" for your shell, try 'key="value"' or key=\"value\""#
```

Comment:
When `--cfg a(b=c)` or similarly malformed argument is provided, such as in the test below, this added note seems really out of place.

Is there a way rustc could detect specifically for the situation where the escapes may be missing? I imagine any case where this situation occurs, rustc will see something like `key=value`.

Alternatively what about if rustc printed out the argument we received and could not parse? Something like

```
expected `key` or `key="value"`, got `key=value`
```

seem like might be just enough information for the user to figure out what may be going wrong?

Reply:
Currently, the value the parser obtained is already in the output:
```
$ rustc --cfg foo=val
error: invalid `--cfg` argument: `foo=val` (expected `key` or `key="value"`)
```
But, this unfortunately doesn't do much in guiding the user towards the escaping issue. If one is to read the message carefully, you can spot the difference of course. But if spotting the difference was easy, the issue wouldn't have been filed. Neither would I have gone through the trouble to incorporate its [suggestion](https://github.com/rust-lang/rust/issues/66450#issuecomment-554457074) to more clearly nudge towards the proper escaping. 

Stating what we received duplicates this information, it does more closely couple the quoted `"value"` next to whatever the user passed in though, so the difference may be easier to spot? It would look like this: 
```
$ ./rustc --cfg foo=val
error: invalid `--cfg` argument: `foo=val` (expected `key` or `key="value"`, got `foo=val`)
```

I definitely agree with you that the current message would look odd in a lot of situations, like when a shell variable substitution goes wrong or something;
```
$ rustc --cfg '$VAR'
error: invalid `--cfg` argument: `$VAR` (expected `key` or `key="value"`, ensure escaping is appropriate for your shell, try 'key="value"' or key=\"value\")
```
Like, in that case it's just the wrong advice, I didn't really think of that failure mode. Basically if any invalid character occurs in the string, we would get this 'hint' now, even though it may not be applicable, this is definitely not desirable.

So thinking about this a bit more, we can easily make the error message more specific by checking for `=` and if that is present check for `="` and ending with `"`. That makes this hint much more specific, only hinting in cases where the user likely tried to do a `key="value"` cfg flag? We will still trigger on the `a(b=c)` situation, but I think that's reasonable? The `(` can't be part of the key, the `)` could be part of the value; `ab="c)"` is allowed.

I've implemented this limiting the scope of printing this hint using the simple rules proposed, see the updated unit tests. Only the `a(b=c)` and (newly added) `key=value` unit tests now print this hint, let me know your thoughts.

Reply:
seems reasonable to me.

cc @estebank may want to take a look here as well.